### PR TITLE
Fix the issue that backup files may leak after cleanup

### PR DESCRIPTION
**Pull request by KanShiori**
### What is changed and how does it work?

1. Add a field `retryCount`, and retry to clean backup files when one cleanup fails
2. Try to clean all files in one cleanup instead of returning when met any errors.
3. Add a field `backoffEnabled` to enable backoff when a deletion API fails. It may be useful when the deletion API is rate limited.

### Code changes

- [x] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [x] Unit test 
- [ ] E2E test 
- [x] Manual test 

1. Create a Backup and add 40w files to the bucket.
2. Delete Backup.
3. Check all files are deleted.

- [ ] No code 

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note
Fix the issue that backup files leak after cleanup
```

### CODE REVIEW

**Review by liubog2008 — APPROVED** [review]
LGTM

File: cmd/backup-manager/app/util/util.go [hunk 1]
```diff
@@ -431,3 +432,22 @@ func GetContextForTerminationSignals(op string) (context.Context, context.Cancel
...
+	var lastErr error
+	for i := 0; i < attempts; i++ {
+		err := fn()
+		if err == nil {
+			return nil
+		}
+
+		lastErr = err
+		if sleep != 0 {
+			time.Sleep(sleep)
```
Comment (csuzhangxc) [inline cmd/backup-manager/app/util/util.go:448]: do we need to support cancel with a context?

Comment (KanShiori) [reply cmd/backup-manager/app/util/util.go:452]: add a callback `retriable func(error) bool` to support cancel, so we can cancel with context in `retriable`.

Comment (csuzhangxc) [reply cmd/backup-manager/app/util/util.go:452]: but we `RetriableOnAnyError` now?

Comment (KanShiori) [reply cmd/backup-manager/app/util/util.go:452]: if the context is done, the delete function will fail fastly, so I think it is ok.

File: tools/tools.go [hunk 2]
```diff
@@ -11,6 +11,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+//go:build tools
```
Comment (DanielZhangQD) [inline tools/tools.go:14]: Remove this?

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 3]
```diff
@@ -1635,6 +1635,12 @@ type CleanOption struct {
 	// PageSize represents the number of objects to clean at a time.
 	// default is 10000
 	PageSize uint64 `json:"pageSize,omitempty"`
+	// RetryCount represents the number of retries in pod when the cleanup fails.
+	// +kubebuilder:default=3
+	RetryCount int `json:"retryCount,omitempty"`
+	// BackoffEnabled represents whether to enable the backoff when a deletion API fails.
+	// It is useful when the deletion API is rate limited.
+	BackoffEnabled bool `json:"backoffEnabled,omitempty"`
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:1643]: Should we enable this by default?

File: pkg/apis/pingcap/v1alpha1/backup.go [hunk 4]
```diff
@@ -30,6 +30,7 @@ var (
 	// defaultCleanOption is default clean option
 	defaultCleanOption = CleanOption{
 		PageSize:          10000,
+		RetryCount:        3,
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/backup.go:33]: Set it to 10 as we want to delete the files as much as possible?

File: cmd/backup-manager/app/util/util.go [hunk 5]
```diff
@@ -431,3 +432,26 @@ func GetContextForTerminationSignals(op string) (context.Context, context.Cancel
...
 	return ctx, cancel
 }
+
+func RetriableOnAnyError(err error) bool {
+	return err != nil
+}
+
+// RetryOnError allows the caller to retry fn in case the error returned by fn.
+// sleep define the interval between two retries.
+func RetryOnError(attempts int, sleep time.Duration, retriable func(error) bool, fn func() error) error {
```
Comment (DanielZhangQD) [inline cmd/backup-manager/app/util/util.go:442]: The `sleep` is always 0?

File: pkg/apis/pingcap/v1alpha1/backup.go (hunk 4 above)
Comment (KanShiori) [reply pkg/apis/pingcap/v1alpha1/backup.go:33]: the backoff of the clean job is 5, so the total retry count is 15, so I think it is enough. And clients can modify the option if needed

File: cmd/backup-manager/app/util/util.go (hunk 1 above)
Comment (KanShiori) [reply cmd/backup-manager/app/util/util.go]: Updated

File: pkg/apis/pingcap/v1alpha1/types.go (hunk 3 above)
Comment (KanShiori) [reply pkg/apis/pingcap/v1alpha1/types.go:1643]: It is useful when API is rate limited, and not suitable for all clients. So I think disable by default.

File: tools/tools.go (hunk 2 above)
Comment (KanShiori) [reply tools/tools.go]: Done

File: pkg/apis/pingcap/v1alpha1/backup.go (hunk 4 above)
Comment (DanielZhangQD) [reply pkg/apis/pingcap/v1alpha1/backup.go:33]: But it's more lightweight to retry in the current Pod instead of waiting for another Pod, right?

Comment (KanShiori) [reply pkg/apis/pingcap/v1alpha1/backup.go]: default to 5

**Review by DanielZhangQD — APPROVED** [review]
LGTM

**Review by csuzhangxc — APPROVED** [review]

**Pull request merged**